// monaco-editor/dev/vs/language/html/monaco.contribution.js
define("vs/language/html/monaco.contribution", ["exports", "../../monaco.contribution-DPdiidye"], (function(exports, src_deprecated_language_html_monaco_contribution_ts) {
  "use strict";
  exports.handlebarDefaults = src_deprecated_language_html_monaco_contribution_ts.handlebarDefaults;
  exports.handlebarLanguageService = src_deprecated_language_html_monaco_contribution_ts.handlebarLanguageService;
  exports.htmlDefaults = src_deprecated_language_html_monaco_contribution_ts.htmlDefaults;
  exports.htmlLanguageService = src_deprecated_language_html_monaco_contribution_ts.htmlLanguageService;
  exports.razorDefaults = src_deprecated_language_html_monaco_contribution_ts.razorDefaults;
  exports.razorLanguageService = src_deprecated_language_html_monaco_contribution_ts.razorLanguageService;
  exports.registerHTMLLanguageService = src_deprecated_language_html_monaco_contribution_ts.registerHTMLLanguageService;
  Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
}));
