// monaco-editor/esm/vs/index.js
import * as index from '../external/monaco-lsp-client/out/index.js';
export { index as lsp };
import './languages/definitions/abap/register.js';
import './languages/definitions/apex/register.js';
import './languages/definitions/azcli/register.js';
import './languages/definitions/bat/register.js';
import './languages/definitions/bicep/register.js';
import './languages/definitions/cameligo/register.js';
import './languages/definitions/clojure/register.js';
import './languages/definitions/coffee/register.js';
import './languages/definitions/cpp/register.js';
import './languages/definitions/csharp/register.js';
import './languages/definitions/csp/register.js';
import './languages/definitions/css/register.js';
import './languages/definitions/cypher/register.js';
import './languages/definitions/dart/register.js';
import './languages/definitions/dockerfile/register.js';
import './languages/definitions/ecl/register.js';
import './languages/definitions/elixir/register.js';
import './languages/definitions/flow9/register.js';
import './languages/definitions/fsharp/register.js';
import './languages/definitions/freemarker2/register.js';
import './languages/definitions/go/register.js';
import './languages/definitions/graphql/register.js';
import './languages/definitions/handlebars/register.js';
import './languages/definitions/hcl/register.js';
import './languages/definitions/html/register.js';
import './languages/definitions/ini/register.js';
import './languages/definitions/java/register.js';
import './languages/definitions/javascript/register.js';
import './languages/definitions/julia/register.js';
import './languages/definitions/kotlin/register.js';
import './languages/definitions/less/register.js';
import './languages/definitions/lexon/register.js';
import './languages/definitions/lua/register.js';
import './languages/definitions/liquid/register.js';
import './languages/definitions/m3/register.js';
import './languages/definitions/markdown/register.js';
import './languages/definitions/mdx/register.js';
import './languages/definitions/mips/register.js';
import './languages/definitions/msdax/register.js';
import './languages/definitions/mysql/register.js';
import './languages/definitions/objective-c/register.js';
import './languages/definitions/pascal/register.js';
import './languages/definitions/pascaligo/register.js';
import './languages/definitions/perl/register.js';
import './languages/definitions/pgsql/register.js';
import './languages/definitions/php/register.js';
import './languages/definitions/pla/register.js';
import './languages/definitions/postiats/register.js';
import './languages/definitions/powerquery/register.js';
import './languages/definitions/powershell/register.js';
import './languages/definitions/protobuf/register.js';
import './languages/definitions/pug/register.js';
import './languages/definitions/python/register.js';
import './languages/definitions/qsharp/register.js';
import './languages/definitions/r/register.js';
import './languages/definitions/razor/register.js';
import './languages/definitions/redis/register.js';
import './languages/definitions/redshift/register.js';
import './languages/definitions/restructuredtext/register.js';
import './languages/definitions/ruby/register.js';
import './languages/definitions/rust/register.js';
import './languages/definitions/sb/register.js';
import './languages/definitions/scala/register.js';
import './languages/definitions/scheme/register.js';
import './languages/definitions/scss/register.js';
import './languages/definitions/shell/register.js';
import './languages/definitions/solidity/register.js';
import './languages/definitions/sophia/register.js';
import './languages/definitions/sparql/register.js';
import './languages/definitions/sql/register.js';
import './languages/definitions/st/register.js';
import './languages/definitions/swift/register.js';
import './languages/definitions/systemverilog/register.js';
import './languages/definitions/tcl/register.js';
import './languages/definitions/twig/register.js';
import './languages/definitions/typescript/register.js';
import './languages/definitions/typespec/register.js';
import './languages/definitions/vb/register.js';
import './languages/definitions/wgsl/register.js';
import './languages/definitions/xml/register.js';
import './languages/definitions/yaml/register.js';
import * as __src_languages_features_css_register_ts from './languages/features/css/register.js';
export { __src_languages_features_css_register_ts as css };
import * as __src_languages_features_html_register_ts from './languages/features/html/register.js';
export { __src_languages_features_html_register_ts as html };
import * as __src_languages_features_json_register_ts from './languages/features/json/register.js';
export { __src_languages_features_json_register_ts as json };
import * as __src_languages_features_typescript_register_ts from './languages/features/typescript/register.js';
export { __src_languages_features_typescript_register_ts as typescript };
import './editor/contrib/anchorSelect/browser/anchorSelect.js';
import './editor/contrib/bracketMatching/browser/bracketMatching.js';
import './editor/contrib/caretOperations/browser/transpose.js';
import './editor/contrib/clipboard/browser/clipboard.js';
import './editor/contrib/codeAction/browser/codeActionContributions.js';
import './editor/browser/widget/codeEditor/codeEditorWidget.js';
import './editor/contrib/codelens/browser/codelensController.js';
import './base/browser/ui/codicons/codicon/codicon.css';
import './editor/contrib/colorPicker/browser/colorPickerContribution.js';
import './editor/contrib/comment/browser/comment.js';
import './editor/contrib/contextmenu/browser/contextmenu.js';
import './editor/contrib/cursorUndo/browser/cursorUndo.js';
import './editor/browser/widget/diffEditor/diffEditor.contribution.js';
import './editor/contrib/diffEditorBreadcrumbs/browser/contribution.js';
import './editor/contrib/dnd/browser/dnd.js';
import './editor/contrib/documentSymbols/browser/documentSymbols.js';
import './editor/contrib/dropOrPasteInto/browser/dropIntoEditorContribution.js';
import './features/find/register.js';
import './editor/contrib/floatingMenu/browser/floatingMenu.contribution.js';
import './editor/contrib/folding/browser/folding.js';
import './editor/contrib/fontZoom/browser/fontZoom.js';
import './editor/contrib/format/browser/formatActions.js';
import './editor/contrib/gotoError/browser/gotoError.js';
import './editor/standalone/browser/quickAccess/standaloneGotoLineQuickAccess.js';
import './editor/contrib/gotoSymbol/browser/link/goToDefinitionAtPosition.js';
import './editor/contrib/gpu/browser/gpuActions.js';
import './editor/contrib/hover/browser/hoverContribution.js';
import './editor/contrib/indentation/browser/indentation.js';
import './editor/contrib/inlayHints/browser/inlayHintsContribution.js';
import './editor/contrib/inlineCompletions/browser/inlineCompletions.contribution.js';
import './editor/contrib/inlineProgress/browser/inlineProgress.js';
import './editor/contrib/inPlaceReplace/browser/inPlaceReplace.js';
import './editor/contrib/insertFinalNewLine/browser/insertFinalNewLine.js';
import './editor/standalone/browser/inspectTokens/inspectTokens.js';
import './editor/standalone/browser/iPadShowKeyboard/iPadShowKeyboard.js';
import './editor/contrib/lineSelection/browser/lineSelection.js';
import './editor/contrib/linesOperations/browser/linesOperations.js';
import './editor/contrib/linkedEditing/browser/linkedEditing.js';
import './editor/contrib/links/browser/links.js';
import './editor/contrib/longLinesHelper/browser/longLinesHelper.js';
import './editor/contrib/middleScroll/browser/middleScroll.contribution.js';
import './editor/contrib/multicursor/browser/multicursor.js';
import './editor/contrib/parameterHints/browser/parameterHints.js';
import './editor/contrib/placeholderText/browser/placeholderText.contribution.js';
import './editor/standalone/browser/quickAccess/standaloneCommandsQuickAccess.js';
import './editor/standalone/browser/quickAccess/standaloneHelpQuickAccess.js';
import './editor/standalone/browser/quickAccess/standaloneGotoSymbolQuickAccess.js';
import './editor/contrib/readOnlyMessage/browser/contribution.js';
import './editor/standalone/browser/referenceSearch/standaloneReferenceSearch.js';
import './editor/contrib/rename/browser/rename.js';
import './editor/contrib/sectionHeaders/browser/sectionHeaders.js';
import './editor/contrib/semanticTokens/browser/viewportSemanticTokens.js';
import './editor/contrib/smartSelect/browser/smartSelect.js';
import './editor/contrib/snippet/browser/snippetController2.js';
import './editor/contrib/stickyScroll/browser/stickyScrollContribution.js';
import './editor/contrib/suggest/browser/suggestInlineCompletions.js';
import './editor/standalone/browser/toggleHighContrast/toggleHighContrast.js';
import './editor/contrib/toggleTabFocusMode/browser/toggleTabFocusMode.js';
import './editor/contrib/tokenization/browser/tokenization.js';
import './editor/contrib/unicodeHighlighter/browser/unicodeHighlighter.js';
import './editor/contrib/unusualLineTerminators/browser/unusualLineTerminators.js';
import './editor/contrib/wordHighlighter/browser/wordHighlighter.js';
import './editor/contrib/wordOperations/browser/wordOperations.js';
import './editor/contrib/wordPartOperations/browser/wordPartOperations.js';
import './editor/browser/coreCommands.js';
import './editor/contrib/caretOperations/browser/caretOperations.js';
import './editor/contrib/dropOrPasteInto/browser/copyPasteContribution.js';
import './editor/contrib/find/browser/findController.js';
import './editor/contrib/gotoSymbol/browser/goToCommands.js';
import './editor/contrib/gotoError/browser/markerSelectionStatus.js';
import './editor/contrib/semanticTokens/browser/documentSemanticTokens.js';
import './editor/contrib/suggest/browser/suggestController.js';
import './editor/common/standaloneStrings.js';
import './base/browser/ui/codicons/codicon/codicon-modifiers.css';
export { CancellationTokenSource, Emitter, KeyCode, KeyMod, MarkerSeverity, MarkerTag, Position, Range, Selection, SelectionDirection, Token, Uri, editor, languages } from './editor/editor.api.js';
