// monaco-editor/dev/vs/language/typescript/monaco.contribution.js
define("vs/language/typescript/monaco.contribution", ["exports", "../../monaco.contribution-Bj7ZDQQk"], (function(exports, src_deprecated_language_typescript_monaco_contribution_ts) {
  "use strict";
  exports.JsxEmit = src_deprecated_language_typescript_monaco_contribution_ts.JsxEmit;
  exports.ModuleKind = src_deprecated_language_typescript_monaco_contribution_ts.ModuleKind;
  exports.ModuleResolutionKind = src_deprecated_language_typescript_monaco_contribution_ts.ModuleResolutionKind;
  exports.NewLineKind = src_deprecated_language_typescript_monaco_contribution_ts.NewLineKind;
  exports.ScriptTarget = src_deprecated_language_typescript_monaco_contribution_ts.ScriptTarget;
  exports.getJavaScriptWorker = src_deprecated_language_typescript_monaco_contribution_ts.getJavaScriptWorker;
  exports.getTypeScriptWorker = src_deprecated_language_typescript_monaco_contribution_ts.getTypeScriptWorker;
  exports.javascriptDefaults = src_deprecated_language_typescript_monaco_contribution_ts.javascriptDefaults;
  exports.typescriptDefaults = src_deprecated_language_typescript_monaco_contribution_ts.typescriptDefaults;
  exports.typescriptVersion = src_deprecated_language_typescript_monaco_contribution_ts.typescriptVersion;
  Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
}));
